### PR TITLE
Add clear method to EntityDocument

### PR DESCRIPTION
**Pull request by jakobw**
By letting entities clear themselves we can get rid of type-specific clearing logic in EditEntity, i.e.
* getting rid of Property specific logic in EditEntity from re-adding the data type
* not resetting the FormId counter for Lexemes

Related patch: https://gerrit.wikimedia.org/r/#/c/429792/

Task on Phabricator: https://phabricator.wikimedia.org/T192264

### CODE REVIEW

**Comment by JeroenDeDauw** [comment]
Funny method. Got depracted, undeprecated, deprecated again, removed, and now added again.

Removal happened in 6.0: https://github.com/wmde/WikibaseDataModel/pull/614

No objection from my side to re-adding it, especially if there is no way to construct a new empty entity of the same type.

**Comment by JeroenDeDauw** [comment]
Two points to consider:

* Assigning new objects to the fields rather than calling `clear` on them breaks references. No idea if this is good or bad for this code
* Maybe a version without mutation (more functional) is possible. Something like `$entity->newWithSameId()` or `$entity->newWithId( $entity->getId() )`

File: src/Entity/Clearable.php [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+/**
+ * Interface for Entity objects that can be cleared.
+ *
+ *
+ * @since 7.4
+ *
+ * @license GPL-2.0+
+ */
+interface Clearable {
```
[the remark this answers is not in the document]
Comment (jakobw) [reply src/Entity/Clearable.php]: ClearableEntity does make it clearer (pun intended). Renamed.
Regarding its use: See the two bullet points and the gerrit patch I mention in the description here. Ideally we're getting rid of instanceof checks. A different question would be whether this should just be part of EntityDocument? Up for debate.

File: tests/unit/Entity/ItemTest.php [hunk 2]
```diff
@@ -742,4 +743,36 @@ public function testGetAliasGroups_sameListAsFingerprint() {
...
+		$clone = $item->copy();
+
+		$item->clear();
+
+		$this->assertEquals( $clone->getId(), $item->getId(), 'cleared Item should keep its id' );
+		$this->assertTrue( $item->isEmpty(), 'cleared Item should be empty' );
+	}
+
+	public function clearableProvider() {
+		return array_merge( $this->instanceProvider(), [
```
[the remark this answers is not in the document]
Comment (JeroenDeDauw) [reply tests/unit/Entity/ItemTest.php]: Some clarification happened on IRC and I do not object to leaving the code as is, though think I'd be cool to ditch the instanceProvider call

**Pull request merged**